components/ee/agent-smith/pkg/agent/agent.go

```diff
@@ -211,12 +213,29 @@ func (agent *Smith) Start(ctx context.Context, callback func(InfringingWorkspace
 		clo = make(chan classifiedProcess, 50)
 	)
 	agent.metrics.RegisterClassificationQueues(cli, clo)
+
+	workspaces := make(map[int]*common.Workspace)
+	wssWriteMutex := &sync.Mutex{}
+
 	defer wg.Wait()
 	for i := 0; i < 25; i++ {
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
 			for i := range cli {
+				// Update the workspaces map if this process belongs to a new workspace
+				//
+				// When agent-smith starts there could be multiple attempts to update
+				// the workspaces map. This is because we do not lock before reading from the map.
+				// So, if multiple processes written to the cli chan are
+				// read in parallel, all of them will get inside the below if check.
+				if _, ok := workspaces[i.Workspace.PID]; !ok {
```

Comment:
How about acquiring the lock before reading from the map? Currently it would be possible that one thread is reading from the map while another is currently modifying the internal data structures of the map.

Reply:
It would not make any difference as the subsequent step will only overwrite the existing entry with the same content. i.e. even if the data in the map is updated we do not care as we just want to make sure that the key exists in the map.


Introducing a lock for reads will block the subsequent classification steps as the number of reads are going to be too many.